### PR TITLE
Implement shadow_price

### LINKED ISSUE #1552: Implement shadow_price and mention dual sign changes in NEWS
https://discourse.julialang.org/t/result-dual-vs-getdual-signs-are-the-opposite/16355

### PR DESCRIPTION
**Pull request by mlubin**
Closes #1552 

@chkwon please check

### CODE REVIEW

File: src/constraints.jl [hunk 1]
```diff
@@ -239,3 +239,98 @@ function set_coefficient(constraint::ConstraintRef{Model, MOICON{F, S}},
         MOI.ScalarCoefficientChange(index(variable), convert(T, value)))
     return
 end
+
+"""
+    shadow_price(constraint::ConstraintRef)
+
+The change in the objective from an infinitesimal relaxation of the constraint.
+This value is computed from [`result_dual`](@ref) and can be queried only when
+`has_result_dual` is `true` and the objective sense is `MinSense` or `MaxSense`.
```
Comment (blegat) [inline src/constraints.jl]: Most user will not understand what the objective sense can be if it is not `MinSense` nor `MaxSense`.

File: test/constraint.jl [hunk 2]
```diff
@@ -284,3 +284,86 @@ end
...
+    variables: x, y
+    maxobjective: 1.0*x
+    xub: x <= 2.0
+    ylb: y >= 0.0
+    c: x + y <= 1.0
+    """,
+    # Optimal duals
+    Dict("xub" => 0.0, "ylb" => 1.0, "c" => -1.0),
+    # Expected shadow prices
+    Dict("xub" => 0.0, "ylb" => 1.0, "c" => 1.0))
```
Comment (blegat) [inline test/constraint.jl]: If we increase the constant of `ylb` by `ε` then, we have in `c` : `1.0 > x + y > x + ε` hence `x < 1 - ε`.
Hence I would expect a shadow price of `-1.0` for `ylb`.

File: src/constraints.jl [hunk 3]
```diff
@@ -239,3 +239,98 @@ function set_coefficient(constraint::ConstraintRef{Model, MOICON{F, S}},
...
+          "of constraint.")
+end
+
+# Internal function.
+function shadow_price_less_than_(dual_value, sense::MOI.OptimizationSense)
+    # When minimizing, the shadow price is nonpositive and when maximizing the
+    # shadow price is nonnegative (because relaxing a constraint can only
+    # improve the objective). By MOI convention, a feasible dual on a LessThan
+    # set is nonpositive, so we flip the sign when maximizing.
+    if sense == MOI.MaxSense
```
Comment (blegat) [inline src/constraints.jl]: Why not merge the two functions into one and do `sense == MOI.MaxSense ⊻ greater_than` ?
One might argue that it decreases readability but the fact that the comment in the other function references this one means that merging them might also increase readability.

Comment (mlubin) [reply src/constraints.jl]: I understand the two cases separately. Also both functions are small enough to fit on a single screen together, so readability seems fine to me.

**Comment by chkwon** [comment]
Thanks. Looks good to me.

**Pull request merged**